### ISSUE TITLE
StreamListener does not create binding's error channel

### ISSUE THREAD

**Issue by olegz** [spring-cloud/spring-cloud-stream#1338]
While discussing the error handling with the user (see https://github.com/spring-cloud/spring-cloud-stream/issues/1333#issuecomment-377875306) it was discovered that when application defines error handler using `@StreamListener` the actual error channel is not created. 

Consider the following:
```
@StreamListener(Sink.INPUT)
public void handle(String value) {
	throw new RuntimeException("BOOM!");
}

@StreamListener("inOne.hello.errors")
public void error(Message<?> message) {
	System.out.println("Handling ERROR: " + message);
}
```
and properties
```
spring.cloud.stream.bindings.input.destination=inOne
spring.cloud.stream.bindings.input.group=hello
```
results in 
```
***************************
APPLICATION FAILED TO START
***************************
Description:
A component required a bean named 'inOne.hello.errors' that could not be found.
```

I was able to reproduce it with current and 1.3 version of SCSt.

The issue was identified in `StreamListenerAnnotationBeanPostProcessor.afterSingletonsInstantiated()` where we presume the channel exists during `applicationContext.getBean(mappedBindingEntry.getKey(), SubscribableChannel.class)` call.

Interestingly enough it does work when such handler defined using one of the SI annotations (e.g., `@ServiceActivator` and where the following code exists and creates such channel
```
catch (DestinationResolutionException e) {
	if (e.getCause() instanceof NoSuchBeanDefinitionException) {
		inputChannel = new DirectChannel();
		this.beanFactory.registerSingleton(inputChannelName, inputChannel);
		inputChannel = (MessageChannel) this.beanFactory.initializeBean(inputChannel, inputChannelName);
	}
        . . .
```

**Comment by garyrussell** [comment]
```java
@StreamListener("inOne.hello.errors")
public void error(Message<?> message) {
	System.out.println("Handling ERROR: " + message);
}
```

This is just wrong.

`@StreamListener` is intended to bind to a channel that is bound to an external destination.

That is not the case for the error channel, it is a local SI error channel so `@ServiceActivator` should be used to consume from it, not `@StreamListener`.

**Comment by olegz** [comment]
@garyrussell So, to sum things up; The `@StreamListener` annotation is intended specifically to define bindings that bridge to _external destinations_. 
Given that destination specific error channel does not have an associated external destination, it is then a prerogative of Spring Integration (SI) and our reliance on it, the handler for such destination must be defined using one of the SI handler annotations (i.e., `@ServiceActivator`, `@Transformer` etc.).

**Comment by olegz** [comment]
Closing it as "won't fix" since it was determined this is not a bug and rather a documentation issue only

**Comment by stavalfi** [comment]
@garyrussell @olegz the error handler (`error` method) is never get called.

Consumer:

```java
@EnableBinding(Sink.class)
@SpringBootApplication
public class SpringCloudStreamKinesisConsumerApplication {

	public static void main(String[] args) {
		SpringApplication.run(SpringCloudStreamKinesisConsumerApplication.class, args);
	}

	@StreamListener(Sink.INPUT) // destination name 'input.myGroup'
	public void handle(String value) {
		throw new RuntimeException("BOOM!");
	}

	@ServiceActivator(inputChannel = "stava2.myGroup.errors")
	public void error(Message<?> message) {
		System.out.println("Handling ERROR: " + message);
	}
}

spring.cloud.stream.bindings.input.destination=stava2
spring.cloud.stream.bindings.input.group=myGroup
```

**Comment by olegz** [comment]
@stavalfi I am not sure how it's possible. I've just taken the above example and tested it with Rabbit Binder, which should not matter, since supported functionality is provided from the stream core. Everything works as expected. However I may need to clarify something.

In the above example you have two error handling methods. This means only one of them will execute and that is `error(..)`. 
```
Handling ERROR: ErrorMessage [payload=org.springframework. . .
```
If you commented out, then the `errorGlobal` will execute.

But you are stating that non of them execute. So I am not sure how that's possible.
Can you please provide a bit more information such as version of SCSt, the binder you're using etc.

**Comment by stavalfi** [comment]
@olegz Thanks for the fast response, and I'm sorry for the delay.

First of all, this is the repository for my consumer & producer (but we are only interested in the consumer).

Secondly, I'm not sure what is `SCSt`.

```java
    <parent>
        <groupId>org.springframework.cloud</groupId>
        <artifactId>spring-cloud-build</artifactId>
        <version>2.0.0.BUILD-SNAPSHOT</version>
    </parent>

        <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-stream-reactive</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-stream</artifactId>
            <version>2.0.0.BUILD-SNAPSHOT</version>
        </dependency>
        <dependency>
            <groupId>com.h2database</groupId>
            <artifactId>h2</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-stream-binder-rabbit</artifactId>
        </dependency>

    <dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>org.springframework.cloud</groupId>
                <artifactId>spring-cloud-dependencies</artifactId>
                <version>${spring-cloud.version}</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
        </dependencies>
    </dependencyManagement>



```

**Comment by olegz** [comment]
Ok, was able to reproduce it with your project. Not sure yet what's going on. Let me work on it.

**Comment by olegz** [comment]
@stavalfi Got it! There was a bug in Rabbit binder related to inconsistencies in recoverer https://github.com/spring-cloud/spring-cloud-stream-binder-rabbit/commit/854e922f325a700d32a18e865dcc26209d5e7409.
As you can see the bug was addressed recently but you are still using RC3 of rabbit-binder.
So, update your POM for consumer and you should be fine. The fix will go into 2.0.0.RELEASE which is tomorrow.
```
<dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-stream-binder-rabbit</artifactId>
            <version>2.0.0.BUILD-SNAPSHOT</version>
</dependency>
```

**Comment by stavalfi** [comment]
@olegz I'm truly amazed by your desire to help! Thank you!

One little note - I also must explicitly override the following artifact's version also: `spring-cloud-stream-binder-rabbit-core` because I guess that `spring-cloud-build` or `spring-cloud-dependencies` gave me that artifact in version `RC3 `:

```java
        <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-stream-binder-rabbit</artifactId>
            <version>2.0.0.BUILD-SNAPSHOT</version>
        </dependency>
        <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-stream-binder-rabbit-core</artifactId>
            <version>2.0.0.BUILD-SNAPSHOT</version>
        </dependency>
```
